apps/evm/src/utilities/requestGaslessTransaction/index.ts

```diff
@@ -0,0 +1,121 @@
+      paymasterParams: {
+        paymaster: `0x${string}`;
+        paymasterInput: `0x${string}`;
+      };
+      gasPerPubdata: number;
+    };
+    maxFeePerGas: number;
+    gasLimit: number;
+  };
+}
+
+export async function requestGaslessTransaction<
+  TContract extends BaseContract,
+  TMethodName extends string & keyof TContract['functions'],
+>(
+  contract: TContract,
+  methodName: TMethodName,
+  args: Omit<Parameters<TContract['functions'][TMethodName]>, 'overrides'>,
+  overrides: { value: string } | Record<never, never> = {},
+): Promise<ContractTransaction> {
+  const chainId = await contract.signer.getChainId();
+  if (chainId === ChainId.ZKSYNC_MAINNET || chainId === ChainId.ZKSYNC_SEPOLIA) {
+    const accountAddress = (await contract.signer.getAddress()) as Address;
+    const txData = {
+      to: contract.address,
+      from: accountAddress,
+      data: contract.interface.encodeFunctionData(methodName, args),
+      ...overrides,
+    };
+
+    const payload = {
+      chainId,
+      sponsorshipRatio: 100,
+      txData,
+    };
+
+    try {
+      const response = await fetch(config.zyFi.sponsoredPaymasterEndpoint, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          'X-API-Key': config.zyFi.apiKey,
+        },
+        body: JSON.stringify(payload),
+      });
+
+      if (!response.ok) {
+        throw new Error(`HTTP error, status: ${response.status}`);
+      }
+
+      const chain = chainId === ChainId.ZKSYNC_MAINNET ? zksync : zksyncSepoliaTestnet;
+
+      const { txData } = (await response.json()) as ZyFiSponsoredTxResponse;
+      const publicClient = createPublicClient({
+        chain,
+        transport: http(config.rpcUrls[chainId]),
+      }).extend(eip712WalletActions());
+
+      const walletClient = createWalletClient({
+        chain,
+        transport: custom(window.ethereum),
+      }).extend(eip712WalletActions());
+
+      const nonce = await publicClient.getTransactionCount({
+        address: accountAddress,
+      });
+      const txPayload = {
+        account: accountAddress,
+        to: txData.to,
+        value: BigInt(txData.value),
+        chain,
+        gas: BigInt(txData.gasLimit),
+        gasPerPubdata: BigInt(txData.customData.gasPerPubdata),
+        maxFeePerGas: BigInt(txData.maxFeePerGas),
+        maxPriorityFeePerGas: BigInt(0),
+        data: txData.data,
+        paymaster: txData.customData.paymasterParams.paymaster,
+        paymasterInput: txData.customData.paymasterParams.paymasterInput,
+        nonce,
+      };
+      const txHash = await walletClient.sendTransaction(txPayload);
+
+      return {
+        hash: txHash,
+        wait: async (confirmations?: number) =>
+          await publicClient.waitForTransactionReceipt({ hash: txHash, confirmations }),
+      };
+    } catch (error) {
+      logError(error);
+      // return a normal transaction
+      return contract.functions[methodName](...args);
```

Comment:
Do you think that the wallet confirmation is not enough then? Because right now, if for whatever reason we can't send a gas-less transaction, you'd get a usual confirmation popup from the wallet UI, so if I understood your idea correctly, we'd have to ask it before even generating the transaction

Reply:
I don't think it is, because the UI would be misleading: we would have told users that transactions are gas-free when they send a transaction (from the "Gas-free" badge in the header), yet when their wallet UI shows up we'd propose to send a paid transaction. We need to be explicit about why that is the case and to make sure not to catch anyone off-guard.

And correct yes, we'll have to ask before generating the transaction; only if the "Gas-free" badge is still showing in the header and after sending a gas-free transaction failed.